ci: Only run-tests if tests/demo/examples/core is modified

angela-ko opened the pull request:
#### Relevant issue or PR
#205 

#### Description of changes
Add filter to only `run-tests` if `tests` `demo` `examples` `tesseract-core` is modified

#### Testing done
This will be the test (if run-tests doesn't trigger)

jpbrodrick89 commented:
Also, should probably consider adding `production.uv.lock` (and maybe `requirements.txt` or `pyproject.toml`) so that they run when dependencies are updated?

dionhaefner commented:
Hmm isn't this pretty much all paths now? 😅 

I suggest a slightly different approach:

- Always run base tests and doc build.
- Always run all tests on pushes to `main`.
- Run `pip install` actions only if `requirements.txt` has changed.
- Run E2E tests and examples only if the following paths have changed:

```yaml
      - "tesseract_core/**"
      - "tests/endtoend_tests/**"
      - "demo/**"
      - "examples/**"
      - "production.uv.lock"
```

(accepting the rare false-negative but that should be OK and immediately caught by the push to main test)

This makes it so we don't need to run the expensive E2E tests if changes are isolated to `docs`, `.github`, `pyproject.toml`, base tests, or `README.md` / other top-level files.

angela-ko commented:
I guess the concern I thought Ale might have (hence why I added) is that if the packages change, the tests may fail due to changes in the packages? But I supposed that would be relatively rare and could be caught in the regular benchmarking? 

I think currently, docs changes + workflow files other than e2e test file changes are excluded, but nothing else. Added base tests to exclusion list is a good catch as well. 

I don't quite understand the production.uv.lock actually -- i thought that's just generated based on the pyproject.toml / requirements? So would adding that be effectively the same as having pyproject.toml in the list?

angela-ko commented:
@dionhaefner I did leave run_tests.yml in there since the tests should be run if the test configs are modified?

angela-ko commented:
> This works for the whole workflow `run_tests`, but if we want to prescribe path for specific **jobs** (i.e.. end-to-end vs base) is that possible to achieve in a similar way?

I saw that there's some hacky way about it using if conditions in the jobs and matching paths? but I think it's best to just split it into 2 workflow files for e2e vs base...

angela-ko commented:
https://github.com/pasteurlabs/tesseract-core/actions/runs/15568981560/job/43840017745?pr=219

Did a test branch here and checked that e2e tests are not fully running when only doc is changed vs in current branch it is fullyg running e2e tests

dionhaefner commented on .github/workflows/run_tests.yml:
```diff
@@ -210,6 +211,21 @@ jobs:
+          # Assume PRs are less than 50 commits
+          fetch-depth: 50
+
+      - name: Get all changed files
```
Should we do this in the `get-e2e-matrix` step instead? Then we can do it once and just skip the entire `tests-e2e` job if nothing has changed.

angela-ko replied:
I thought about that but `tests-demo` doesn't use `get-e2e-matrix`...

angela-ko commented on .github/workflows/run_tests.yml:
```diff
@@ -210,6 +211,21 @@ jobs:
       - name: Set up Git repository
         uses: actions/checkout@v4
+        with:
+          # Assume PRs are less than 50 commits
+          fetch-depth: 50
```
haha all the real life examples provided in the repo used `fetch-depth: 50` i'm guessing because of the size of the repos? but I can change to 0 if it's actually just like 2 seconds extra

dionhaefner replied:
Yes, let's use `0` and revert if this ever becomes a measurable overhead.

angela-ko commented:
https://github.com/pasteurlabs/tesseract-core/actions/runs/15573704318/job/43854596186?pr=219

Ok reran the tests and now it's fully showing up as skipped (did not know if conditions could be put on the full job)

<img width="589" alt="image" src="https://github.com/user-attachments/assets/ba64db7d-ac23-43d7-ab80-0a5b19174076" />

vs tests are running here
